### ISSUE TITLE
EventEmitter Memory Leak Warning

### ISSUE THREAD

**Issue by gevalo1** [newrelic/node-newrelic#463]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT share sensitive information, whether personal, proprietary, or otherwise! )

We see the following warnings when running our tests with Jest, we haven't seen it while running our application yet.
It doesn't happen when we only run a few tests, but only when we run all of our test suites.

It's interesting I only get this issue on `Debian 9.5` and a `node:9-alpine` docker container but not on my local machine running Windows 10 (running in git for Windows).

```
(node:4738) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 unpipe listeners added. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:256:17)
    at WriteStream.addListener (events.js:272:10)
    at WriteStream.Readable.on (_stream_readable.js:843:35)
    at Logger.Object.<anonymous>.Readable.pipe (/var/www/hidden/node_modules/newrelic/node_modules/readable-stream/lib/_stream_readable.js:631:8)
    at new Logger (/var/www/hidden/node_modules/newrelic/lib/util/logger.js:46:10)
    at Object.<anonymous> (/var/www/hidden/node_modules/newrelic/lib/logger.js:7:18)
    at Runtime._execModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:1205:24)
    at Runtime._loadModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:805:12)
    at Runtime.requireModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:662:10)
    at Runtime.requireModuleOrMock (/var/www/hidden/node_modules/jest-runtime/build/index.js:826:21)
(node:4738) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 drain listeners added. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:256:17)
    at WriteStream.addListener (events.js:272:10)
    at WriteStream.Readable.on (_stream_readable.js:843:35)
    at Logger.Object.<anonymous>.Readable.pipe (/var/www/hidden/node_modules/newrelic/node_modules/readable-stream/lib/_stream_readable.js:654:8)
    at new Logger (/var/www/hidden/node_modules/newrelic/lib/util/logger.js:46:10)
    at Object.<anonymous> (/var/www/hidden/node_modules/newrelic/lib/logger.js:7:18)
    at Runtime._execModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:1205:24)
    at Runtime._loadModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:805:12)
    at Runtime.requireModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:662:10)
    at Runtime.requireModuleOrMock (/var/www/hidden/node_modules/jest-runtime/build/index.js:826:21)
(node:4738) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:256:17)
    at WriteStream.prependListener (events.js:279:14)
    at prependListener (/var/www/hidden/node_modules/newrelic/node_modules/readable-stream/lib/_stream_readable.js:98:69)
    at Logger.Object.<anonymous>.Readable.pipe (/var/www/hidden/node_modules/newrelic/node_modules/readable-stream/lib/_stream_readable.js:708:3)
    at new Logger (/var/www/hidden/node_modules/newrelic/lib/util/logger.js:46:10)
    at Object.<anonymous> (/var/www/hidden/node_modules/newrelic/lib/logger.js:7:18)
    at Runtime._execModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:1205:24)
    at Runtime._loadModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:805:12)
    at Runtime.requireModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:662:10)
    at Runtime.requireModuleOrMock (/var/www/hidden/node_modules/jest-runtime/build/index.js:826:21)
(node:4738) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:256:17)
    at WriteStream.addListener (events.js:272:10)
    at WriteStream.Readable.on (_stream_readable.js:843:35)
    at WriteStream.once (events.js:301:8)
    at Logger.Object.<anonymous>.Readable.pipe (/var/www/hidden/node_modules/newrelic/node_modules/readable-stream/lib/_stream_readable.js:715:8)
    at new Logger (/var/www/hidden/node_modules/newrelic/lib/util/logger.js:46:10)
    at Object.<anonymous> (/var/www/hidden/node_modules/newrelic/lib/logger.js:7:18)
    at Runtime._execModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:1205:24)
    at Runtime._loadModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:805:12)
    at Runtime.requireModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:662:10)
(node:4738) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 finish listeners added. Use emitter.setMaxListeners() to increase limit
    at _addListener (events.js:256:17)
    at WriteStream.addListener (events.js:272:10)
    at WriteStream.Readable.on (_stream_readable.js:843:35)
    at WriteStream.once (events.js:301:8)
    at Logger.Object.<anonymous>.Readable.pipe (/var/www/hidden/node_modules/newrelic/node_modules/readable-stream/lib/_stream_readable.js:723:8)
    at new Logger (/var/www/hidden/node_modules/newrelic/lib/util/logger.js:46:10)
    at Object.<anonymous> (/var/www/hidden/node_modules/newrelic/lib/logger.js:7:18)
    at Runtime._execModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:1205:24)
    at Runtime._loadModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:805:12)
    at Runtime.requireModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:662:10)
```

## Expected Behavior

[NOTE]: # ( Tell us what you expected to happen. )

The listener limit shouldn't be reached. I'm also not sure why this is even happening while running tests?

## Steps to Reproduce

[NOTE]: # ( Please be as specific as possible. )
[TIP]:  # ( Link a sample application that demonstrates the issue. )

Not entirely sure yet, we just use Jest to run our tests but don't have any New Relic specific code or do anything at all custom.
Just running a lot of tests seems to trigger it for us.

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )

Node version: Both 9.11.2 & 10.21.0
OS: Both Alpine Linux & Debian 9.5, but not Windows 10
Agent Version: 5.13.1

**Comment by carlonewrelic** [comment]
Hi @gevalo1 , nice to meet you and thank you for bringing up this issue with us!  I just have a few questions.

Are you seeing this issue occur in both Node 9 and 10?  In the issue description you mention Node 9.

Does the max listeners issue occur with the New Relic agent disabled?

What version of Jest are you using?  It seems like there's some sort of build step happening(?)
`at Runtime._execModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:1205:24)`

It appears that the memory leak is happening with the agent attempts to log.  
`at new Logger (/var/www/hidden/node_modules/newrelic/lib/util/logger.js:46:10)`
Can you try disabling logging to see if that clears up the issue?  You can do that by setting this value in the newrelic config file:
```
logging: {
  enabled: false
}
```
OR
setting this environment variable to false:  `NEW_RELIC_LOG_ENABLED`

**Comment by gevalo1** [comment]
Hi @carlonewrelic,

This issue does indeed occur using both Node 9 & 10, haven't tested any other versions.

The issue still occurs even if I change the configuration & environment variables to disable both the agent itself & the logging.

We are currently using Jest version 25.5.4.

**Comment by carlonewrelic** [comment]
Thank you for trying out those things @gevalo1.  In order to better reproduce what you're seeing can you give me any general details about the node app you are testing?  Could you provide what module dependencies you have and provide a general outline of the apps functionality?  Is there a lot of async code in the application?  Does it make database or external calls?  Any other details your can provide would be extremely helpful.

**Comment by carlonewrelic** [comment]
@gevalo1 I misread your last post and apologize for that.  Just to clarify, you are seeing the error occur with the agent disabled as well?  Did you disable the agent by removing `require('newrelic')` from your app or via config `agent_enabled` (`NEW_RELIC_ENABLED` env var)?

**Comment by gevalo1** [comment]
I disabled it using the config & environment variable, neither of these helped. However I just tried removing `require('newrelic')` and then the warnings are gone. But this is of course not a permanent solution.

We do use an in-memory sqlite3 database in our tests & there is also quite a bit of async code in the application.

For sqlite3 we use version 4.0.9, but more specifically we use this branch: https://github.com/rlindner81/node-sqlite3/tree/user-functions.

For Jest we also have a few custom [matchers](https://jestjs.io/docs/en/expect#custom-matchers-api) & [reporters](https://jestjs.io/docs/en/configuration#reporters-arraymodulename--modulename-options), another package we use related to this is `jest-diff`.

**Comment by carlonewrelic** [comment]
Thank you @gevalo1 for providing the extra details on the issue.  We'll try and reproduce what you're seeing and will reach out to you if we need more information.

**Comment by astormnewrelic** [comment]
@gevalo1 we took a first pass at this and weren't able to reproduce it. 

Can you talk a bit about how your jest tests work or how they're structured or point us towards an isolated reproduction?  

TL;DR; -- this is likely a case where we'll need to see the issue happen in order to diagnose what's happening.  An error callstack by itself (while useful) won't be enough.  The issue appears to be some poor interaction between jest and the agnet's logger.  The rest of this message describes how our logger works and why we think jest might be at the heart of the issue.  If anything in here triggers recognition plase let us know. 

## The Node Agent's Logger

The Agent's [underlying logger object](https://github.com/newrelic/node-newrelic/blob/16f99e62a229d7b8f73561c7df3bfaf167fd212a/lib/util/logger.js) implements [node's _readable stream_ interface](https://github.com/newrelic/node-newrelic/blob/16f99e62a229d7b8f73561c7df3bfaf167fd212a/lib/util/logger.js#L31).  

When a client-programmer logs a value, we add data to the object's `buffer` property. (this is an implementation detail, and not part of the readable streams interface).  

Then, in `_read`, (which IS part of the readable stream interface), we clear this buffer and `push` data to the stream.  Node.js will automatically call `_read` when using the logger as a readable stream.

In order to _see_ the data being logged, we need to setup a _writable_ stream, and then pipe the contents of this readable stream to the writable sterem.  We do that in [`lib/logger.js`](https://github.com/newrelic/node-newrelic/blob/main/lib/logger.js).  Specifically, depending [on the logging configuration](https://github.com/newrelic/node-newrelic/blob/16f99e62a229d7b8f73561c7df3bfaf167fd212a/lib/logger.js#L36), we create a stream that writes to either standard out, standard error, or to a file.  

Once the writable stream is created, we pass that writable stream [to the `pipe`](https://github.com/newrelic/node-newrelic/blob/16f99e62a229d7b8f73561c7df3bfaf167fd212a/lib/logger.js#L55) method of the Logger object.  When you `pipe` a writable stream to a readable stream like this, you're asking that readable stream to send its data to the writable stream.  It's this action that makes Node.js call `_read` on the readable stream.

## Too Many Listeners Added 

Looking at the error callstack, the _too many listeners_ added errors

```
11 unpipe listeners added
11 drain listeners added
11 error listeners added
11 close listeners added
11 finish listeners added
```

all seem to be hapeen when we ask Node.js to pipe the writable stream to our Readable stream.  This shouldn't be possible during normal operation of the Agent.  Because Node.js caches the call to each `require` statment our logger instantiation (and it's call to `pipe`) should only happen _once_ during the normal execution of a program.  It also looks like the first item in the call stacks trace back to jest

```
at Runtime.requireModule (/var/www/hidden/node_modules/jest-runtime/build/index.js:662:10)
Runtime.requireModuleOrMock (/var/www/hidden/node_modules/jest-runtime/build/index.js:826:21)
```

Facebook only publishes the original TypeScript of these files to GitHub and not the built source (i.e. these line number won't be 100% accurate to these links), but it looks like [both](https://github.com/facebook/jest/blob/v25.5.4/packages/jest-runtime/src/index.ts#L662) of [these](https://github.com/facebook/jest/blob/v25.5.4/packages/jest-runtime/src/index.ts#L826) are near code where jest is loading modules.  

Our best guess is that jest is doing some non-standard stuff that's not-honoring/partially-honoring the usual contract with `require` to only load a CommonJS module once -- but it's only a guess and hard to say without seeing the problematic behavior. 

@gevalo1 if you (or anyone else watching) can provide an isolated example of this behavior we'll be able to investigate further.  Without it, we're not likely yo get to the bottom of this as none of us are jest experts.

**Comment by astormnewrelic** [comment]
@gevalo1 👋 No rush, but we just wanted to check in to see if you could provide the information we've asked for above.  We're basically looking for information on how you're using New Relic and jest together, as jest and the agent didn't seem compatible when we tried running their initial naive "hello world" examples.  If you can provide us with information on how you run a basic test with jest that might point to the problem here.

**Comment by gevalo1** [comment]
Hi @astormnewrelic 

Sorry for the delay, unfortunately I haven't found the time to look into this further yet but I'll try provide as much information as I can.

We have our own custom package to build up our NodeJS applications, this package contains all kinds of utilities and functions that can be used within those NodeJS applications. At the very top of the `index.js` file of this package we have the `import 'newrelic';` statement, this gets compiled to `require('newrelic');` at build-time.

We currently don't use the Agent API and so we don't import newrelic anywhere else.

We also don't run tests on our `index.js` file so I'm honestly not sure how newrelic ends up being invoked (I'm not super familiar with Jest's internals either).

When commenting out `require('newrelic')` in `node_modules/__our_custom_package__/lib/index.js` then the leaks are gone. Our tests also seem to run 5-10 seconds faster when the newrelic import is removed.

**Comment by astormnewrelic** [comment]
Thanks @gevalo1 -- that's useful information.  Also -- no rushing intended :)  We definitely know what busy schedules look like 

Based on what you've said -- we have two quick follow up question.  

1. Do you know what you're using to compile your ES6 imports into CommonJS require statements? Webpack, Babel, something else?

2. You mention removing the New Relic `require` from `node_modules/__our_custom_package__/lib/index.js` -- this implies that you're somehow running tests in this package from a main package?  Is this correct?  Regardless of that -- do you know what the jest comment you invoke to run your tests looks like and what your jest config looks like?

3. Our basic goal is to be able to reproduce the error you're seeing on our own computers, so if there's anything additional you can think of that would help towards that goal please let us know.

**Comment by gevalo1** [comment]
Hi @astormnewrelic,

1. We use babel.

2. We aren't running tests in node_modules, we only run tests in the main application. In this case only tests inside the current `src` folder are being run.

Since we use a base package which contains several of our scripts (like `test`) we use `execa` to run the command. See https://github.com/sindresorhus/execa#execafile-arguments-options.

The parameters we pass to this function are: `.\node_modules\jest\bin\jest.js [ '--rootDir=C:\\...\\project-name' ] { stdio: 'inherit' }`

Our Jest config looks like this:

```
const path = require('path');

const jest_config = { ...require('base-package/lib/config/jest') };

jest_config.reporters.push(
  [
    require.resolve('./../a-custom-reporter'),
    {
      outputFile: path.resolve(__dirname, 'abc', 'xyz.json'),
    },
  ],
  [
    require.resolve('./../another-custom-reporter'),
    {
      outputFile: path.resolve(__dirname, 'abc', 'def.json'),
    },
  ]
);

jest_config.setupFilesAfterEnv = [require.resolve('./jest/setup')]; // This includes our custom Jest matchers

module.exports = jest_config;
```

The Jest config in the "base-package" looks like this:

```
{
  verbose: false,
  globalSetup: require.resolve('./...'),
  globalTeardown: require.resolve('./...'),
  roots: [...],
  testPathIgnorePatterns: ['/node_modules/', `${outputDirectory()}`],
  testURL: 'https://localhost',
  reporters: [
    'default',
    [
      require.resolve('./...'),
    ],
  ],
  collectCoverageFrom: [
    `${sourceDirectory()}/**/*.{js,jsx,mjs}`,
    'test/utils/**/*.{js,jsx,mjs}',
    ...,
  ].filter(Boolean),
  coverageDirectory: './coverage/',
}
```

3. As mentioned, I don't seem to encounter the issue on Windows, so it might not occur at all in all environments. The OS where I get the issue is Debian 9.

If I think of anything else I'll definitely share it!

**Comment by MathieuTricoire** [comment]
Hi,
I ran into the same issue as @gevalo1 on one project I’m working on, I dug a bit and came to the same conclusion about the logger, this happens because a logger is created each time the module is loaded (side-effect), I checked inside Jest and saw that modules are loaded for each test suites (isolation).

There is a simple repo to reproduce the error:
https://github.com/MathieuTricoire/jest-newrelic-memory-leak
Install the project and run `npm test`

If you comment `require(‘newrelic’)` the leak disappear.
Do you guys have an idea how to fix that? One way would be to give an access to the developer to teardown the logger but I don’t know if it’s the most practical solution…

And today I found that people already talk about these memory leaks problems on Jest:
 https://chanind.github.io/javascript/2019/10/12/jest-tests-memory-leak.html
https://github.com/facebook/jest/issues/6814

**Comment by nijotz** [comment]
Hi Mathieu, thanks for the repro and the info! That helps a lot. One thing that stands out to me is that the article you linked shows how the sentry module wraps the https.request function for analytics. The newrelic module does the same thing. It seems like to prevent the memory leak, we'd have to not wrap anything in the http module. As the article states:
>All it takes to trigger your tests to leak is to require a third-party module that decorates a native module anywhere in your code. Error logging / analytics modules tend to do this sort of decorating of native modules, but theoretically any module can do it.

One thing I've seen people do is put their `require('newrelic')` in an `if` statement that checks if the environment is production. That's an option to avoid this memory leak during test runs.

Also [this comment](https://github.com/facebook/jest/issues/6814#issuecomment-718824886) seems to have a way around it, where a new process is used for each suite.

It appears that we'd have to make significant changes to the agent architecture to avoid this leak. Or detect the use of jest and not wrap native modules. But that's pretty much the same as only requiring newrelic in production environments. I suppose there is some value in testing with the newrelic module required (like being able to catch bugs that might pop up in production due to use of the newrelic module) but it seems like a decent work around would be not use newrelic when using jest, or try the workaround that was left in the github issue.

Also it's the end of my workday and I wanted to get my thoughts out there before they were lost to sleep. There may be another approach I'm missing or some inaccurate info, I'll double check tomorrow when I have so more time.

**Comment by MathieuTricoire** [comment]
I agree with you @nijotz 👍
On my side I created a lib that delegates to the new relic's lib in production and do nothing otherwise.

**Comment by nijotz** [comment]
Seems like a jest issue, nothing we can do here on the agent side, a workaround exists. Closing
